cfpb/cfgov-refresh
# Add global banner manager to Wagtail

We have created a lot of banners that live above everything else, at the very top of the page, with a [complex](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/organisms/header.html#L28-L61) [flag-based](https://github.com/cfpb/cfgov-refresh/blob/6de3110eac363de2969043fbe7c8ccb51f50632b/cfgov/cfgov/settings/base.py#L631-L656) [apparatus](https://github.com/search?q=repo%3Acfpb%2Fcfgov-refresh+extension%3Ahtml+filename%3Abanner+language%3AHTML+language%3AHTML&type=Code&ref=advsearch&l=HTML&l=HTML). This requires developer intervention for every new banner that gets created, and it is piling on a lot of conditional code and tech debt that could better be handled by storing some of this configuration in the database.

This PR creates a new Wagtail-based banner manager to do just that. A new Banner Django model is added, along with a ModelAdmin interface for creating and editing Banner instances.

Banners have a `title` used for internal reference, a `url_patterns` field for determining what paths it appears on, an instance of the `Notification` molecule (providing the visible content of the notification), and an `enabled` checkbox to manually enable/disable the banner.

See: https://GHE/CFGOV/platform/issues/3677


## Additions

- `Banner` model in the `v1` app
- `banners` context variable on `CFGOVPage` for templates to check to see if any banners apply to the given request
- Template code in `organisms/header.html` to check said `banners` variable and output any banners that are in it


## Testing

1. Pull branch
1. Migrate
1. Go to http://localhost:8000/admin/v1/banner/ and create a new banner
1. Check that it appears on the pages you expect, and doesn't on the pages you don't expect


## Screenshots

### Banner editor

![](https://user-images.githubusercontent.com/1044670/77700369-722dde00-6f8a-11ea-9a00-51474e795798.png)

### Result on page

![](https://user-images.githubusercontent.com/1044670/77700472-a0abb900-6f8a-11ea-82c7-acf4485facd3.png)



## Notes

- I don't immediately see why my additions to the `CFGOVPage` `get_context` function are causing errors like this with existing tests:

```
======================================================================
ERROR: test_get_context_calls_get_context (v1.tests.models.test_base.TestCFGOVPage)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/cranfills/Projects/cfgov-refresh/cfgov/v1/tests/models/test_base.py", line 32, in test_get_context_calls_get_context
    self.page.get_context(self.request)
  File "/Users/cranfills/Projects/cfgov-refresh/cfgov/v1/models/base.py", line 214, in get_context
    enabled_banners = Banner.objects.filter(enabled=True)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/query.py", line 787, in filter
    return self._filter_or_exclude(False, *args, **kwargs)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/query.py", line 805, in _filter_or_exclude
    clone.query.add_q(Q(*args, **kwargs))
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1250, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1260, in _add_q
    connector = q_object.connector
AttributeError: 'Q' object has no attribute 'connector'
```

## Checklist
